server/utils/passport/local-signup.js

```diff
@@ -1,91 +1,56 @@
-      const username = reqBody.username
-      const email = reqBody.email
-      const display_name = reqBody.display_name
-      const { appDb } = req.app.locals
-      const hashedPW = hash(password)
-      appDb.collection(collections.users).findOneAndUpdate(
-        { uid: username },
-        {
-          $set: {
-            display_name: display_name,
-            title: '',
-            department: '',
-            company: '',
-            mail: email,
-            member_of: '',
-            bad_pwd_count: 0,
-            lockout_time: 0,
-            last_logoff: Date.now(),
-            last_logon: Date.now(),
-            account_expires: null,
-            password: hashedPW,
-            ldap: false,
-            force_reset_pw: false,
-            realms: config.app.realms,
-            icon: '',
-            access: [],
-            blocked: false,
-            role: 'member',
-          },
-        },
-        {
-          projection: {
-            _id: 0,
-            uid: 1,
-            display_name: 1,
-            acl: 1,
-            role: 1,
-            icon: 1,
-            mail: 1,
-          },
-          returnOriginal: false,
-          upsert: true,
-        },
-        function (err, userinfo) {
-          if (err) {
-            console.error(err.message)
-          }
+    async function (err, existingUser, reqBody) {
+      try {
+        if (err) return res.redirect(`${routes.login}?e=${err}`)
+
+        if (existingUser) return res.redirect(`${routes.signup}?e=existingUser`)
 
-          var defaultConfig = {
-            owner: username,
-            config: defaultUserConfig,
-            name: 'default',
-            type: 'matrix',
-            readers: [username],
-            created: new Date().toUTCString(),
-          }
-          appDb.collection(collections.configs).insertOne(defaultConfig)
+        const { appDb } = req.app.locals
+        const password = reqBody.password
+        const uid = reqBody.username
+        const email = reqBody.email
+        const display_name = reqBody.display_name
+        const hashedPW = hash(password)
+        const account_expires = moment().add(2, 'years').format()
+        const config = ConfigModel.default({ owner: uid, readers: [uid] })
+        const newUser = UserModel.default({
+          uid,
+          display_name,
+          password: hashedPW,
+          mail: email,
+          account_expires,
+        })
 
-          req.session.role = userinfo.value.role
-          req.session.uid = userinfo.value.uid
-          req.session.display_name = userinfo.value.display_name
-          req.session.mail = userinfo.value.mail
-          req.session.celery_tasks = []
-          req.session.icon = userinfo.value.icon
-          return res.redirect(`${basePath}/`)
-        }
-      )
+        await appDb.collection(collections.users).insertOne(newUser)
+        await appDb.collection(collections.configs).insertOne(config)
```

Comment:
If you use the `save` methods on the models you could remove the `default` calls above.

```suggestion
        await UserModel.save(appDb, { uid, display_name, password: hashedPW, mail: email, account_expires })
        await ConfigModel.save(appDb, { owner: uid, readers: [uid] })
```

Reply:
[88c0683](https://github.com/AMP-SCZ/dpdash/pull/293/commits/88c0683e88c4bba873baf2a78d32f796c0c889a5)